## Cryptocurrency

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

#### Load Data

In [2]:
crypto_data = pd.read_csv('./Resources/crypto_data.csv')

In [3]:
crypto_data

Unnamed: 0        CoinName Algorithm  IsTrading ProofType  \
0            42         42 Coin    Scrypt       True   PoW/PoS   
1           365         365Coin       X11       True   PoW/PoS   
2           404         404Coin    Scrypt       True   PoW/PoS   
3           611       SixEleven   SHA-256       True       PoW   
4           808             808   SHA-256       True   PoW/PoS   
...         ...             ...       ...        ...       ...   
1247        XBC     BitcoinPlus    Scrypt       True       PoS   
1248       DVTC      DivotyCoin    Scrypt      False   PoW/PoS   
1249       GIOT     Giotto Coin    Scrypt      False   PoW/PoS   
1250       OPSC  OpenSourceCoin   SHA-256      False   PoW/PoS   
1251       PUNK       SteamPunk       PoS      False       PoS   

      TotalCoinsMined TotalCoinSupply  
0        4.199995e+01              42  
1                 NaN      2300000000  
2        1.055185e+09       532000000  
3                 NaN          611000  
4        0.000000e+00               0  
...               ...             ...  
1247     1.283270e+05         1000000  
1248     2.149121e+07       100000000  
1249              NaN       233100000  
1250              NaN        21000000  
1251              NaN        40000000  

[1252 rows x 7 columns]

#### Data Preparation

Removing cryptocurrency that is not currently being traded.

In [4]:
crypto_1 = crypto_data.loc[(crypto_data["IsTrading"]==True)]
crypto_1

Unnamed: 0     CoinName    Algorithm  IsTrading ProofType  \
0            42      42 Coin       Scrypt       True   PoW/PoS   
1           365      365Coin          X11       True   PoW/PoS   
2           404      404Coin       Scrypt       True   PoW/PoS   
3           611    SixEleven      SHA-256       True       PoW   
4           808          808      SHA-256       True   PoW/PoS   
...         ...          ...          ...        ...       ...   
1243       SERO   Super Zero       Ethash       True       PoW   
1244        UOS          UOS      SHA-256       True      DPoI   
1245        BDX       Beldex  CryptoNight       True       PoW   
1246        ZEN      Horizen     Equihash       True       PoW   
1247        XBC  BitcoinPlus       Scrypt       True       PoS   

      TotalCoinsMined TotalCoinSupply  
0        4.199995e+01              42  
1                 NaN      2300000000  
2        1.055185e+09       532000000  
3                 NaN          611000  
4        0.000000e+00               0  
...               ...             ...  
1243              NaN      1000000000  
1244              NaN      1000000000  
1245     9.802226e+08      1400222610  
1246     7.296538e+06        21000000  
1247     1.283270e+05         1000000  

[1144 rows x 7 columns]

In [5]:
crypto_2 = crypto_1.drop(columns=['IsTrading'])
crypto_2

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
1           365      365Coin          X11   PoW/PoS              NaN   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
3           611    SixEleven      SHA-256       PoW              NaN   
4           808          808      SHA-256   PoW/PoS     0.000000e+00   
...         ...          ...          ...       ...              ...   
1243       SERO   Super Zero       Ethash       PoW              NaN   
1244        UOS          UOS      SHA-256      DPoI              NaN   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
1         2300000000  
2          532000000  
3             611000  
4                  0  
...              ...  
1243      1000000000  
1244      1000000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[1144 rows x 6 columns]

Droping rows with at least one missing element.

In [6]:
crypto_3 = crypto_2.dropna()
crypto_3

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
4           808          808      SHA-256   PoW/PoS     0.000000e+00   
5          1337    EliteCoin          X13   PoW/PoS     2.927942e+10   
7           BTC      Bitcoin      SHA-256       PoW     1.792718e+07   
...         ...          ...          ...       ...              ...   
1238       ZEPH       ZEPHYR      SHA-256      DPoS     2.000000e+09   
1242        GAP      Gapcoin       Scrypt   PoW/PoS     1.493105e+07   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
2          532000000  
4                  0  
5       314159265359  
7           21000000  
...              ...  
1238      2000000000  
1242       250000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[685 rows x 6 columns]

Removing Cryptocurrency that has not been mined.

In [7]:
crypto_4 = crypto_3.loc[(crypto_3["TotalCoinsMined"]>0)]
crypto_4

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
5          1337    EliteCoin          X13   PoW/PoS     2.927942e+10   
7           BTC      Bitcoin      SHA-256       PoW     1.792718e+07   
8           ETH     Ethereum       Ethash       PoW     1.076842e+08   
...         ...          ...          ...       ...              ...   
1238       ZEPH       ZEPHYR      SHA-256      DPoS     2.000000e+09   
1242        GAP      Gapcoin       Scrypt   PoW/PoS     1.493105e+07   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
2          532000000  
5       314159265359  
7           21000000  
8                  0  
...              ...  
1238      2000000000  
1242       250000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[532 rows x 6 columns]

Deleting CoinName as the data must be numeric to be comprehensible to a machine learning algorithm and it does not contribute to analysis.

In [8]:
crypto_5 = crypto_4.drop(columns=['Unnamed: 0','CoinName'])
crypto_5

Algorithm ProofType  TotalCoinsMined TotalCoinSupply
0          Scrypt   PoW/PoS     4.199995e+01              42
2          Scrypt   PoW/PoS     1.055185e+09       532000000
5             X13   PoW/PoS     2.927942e+10    314159265359
7         SHA-256       PoW     1.792718e+07        21000000
8          Ethash       PoW     1.076842e+08               0
...           ...       ...              ...             ...
1238      SHA-256      DPoS     2.000000e+09      2000000000
1242       Scrypt   PoW/PoS     1.493105e+07       250000000
1245  CryptoNight       PoW     9.802226e+08      1400222610
1246     Equihash       PoW     7.296538e+06        21000000
1247       Scrypt       PoS     1.283270e+05         1000000

[532 rows x 4 columns]

Using get_dummies() to convert features with text to numeric values. We have 5 columns and 532 rows.

In [9]:
crypto_6 = pd.get_dummies(crypto_5, columns=["Algorithm", "ProofType"])
crypto_6

TotalCoinsMined TotalCoinSupply  Algorithm_1GB AES Pattern Search  \
0        4.199995e+01              42                                 0   
2        1.055185e+09       532000000                                 0   
5        2.927942e+10    314159265359                                 0   
7        1.792718e+07        21000000                                 0   
8        1.076842e+08               0                                 0   
...               ...             ...                               ...   
1238     2.000000e+09      2000000000                                 0   
1242     1.493105e+07       250000000                                 0   
1245     9.802226e+08      1400222610                                 0   
1246     7.296538e+06        21000000                                 0   
1247     1.283270e+05         1000000                                 0   

      Algorithm_536  Algorithm_Argon2d  Algorithm_BLAKE256  Algorithm_Blake  \
0                 0                  0                   0                0   
2                 0                  0                   0                0   
5                 0                  0                   0                0   
7                 0                  0                   0                0   
8                 0                  0                   0                0   
...             ...                ...                 ...              ...   
1238              0                  0                   0                0   
1242              0                  0                   0                0   
1245              0                  0                   0                0   
1246              0                  0                   0                0   
1247              0                  0                   0                0   

      Algorithm_Blake2S  Algorithm_Blake2b  Algorithm_C11  ...  \
0                     0                  0              0  ...   
2                     0                  0              0  ...   
5                     0                  0              0  ...   
7                     0                  0              0  ...   
8                     0                  0              0  ...   
...                 ...                ...            ...  ...   
1238                  0                  0              0  ...   
1242                  0                  0              0  ...   
1245                  0                  0              0  ...   
1246                  0                  0              0  ...   
1247                  0                  0              0  ...   

      ProofType_PoW/PoS  ProofType_PoW/PoS   ProofType_PoW/PoW  \
0                     1                   0                  0   
2                     1                   0                  0   
5                     1                   0                  0   
7                     0                   0                  0   
8                     0                   0                  0   
...                 ...                 ...                ...   
1238                  0                   0                  0   
1242                  1                   0                  0   
1245                  0                   0                  0   
1246                  0                   0                  0   
1247                  0                   0                  0   

      ProofType_PoW/nPoS  ProofType_Pos  ProofType_Proof of Authority  \
0                      0              0                             0   
2                      0              0                             0   
5                      0              0                             0   
7                      0              0                             0   
8                      0              0                             0   
...                  ...            ...                           ...   
1238                   0              0                    

Wow! The number of columns increased by 94!

Standardize the dataset using StandardScaler to scale the data to have a mean of 0 and a variance of 1.

In [10]:
crypto_6.columns

Index(['TotalCoinsMined', 'TotalCoinSupply',
       'Algorithm_1GB AES Pattern Search', 'Algorithm_536',
       'Algorithm_Argon2d', 'Algorithm_BLAKE256', 'Algorithm_Blake',
       'Algorithm_Blake2S', 'Algorithm_Blake2b', 'Algorithm_C11',
       'Algorithm_Cloverhash', 'Algorithm_Counterparty',
       'Algorithm_CryptoNight', 'Algorithm_CryptoNight Heavy',
       'Algorithm_CryptoNight-V7', 'Algorithm_Cryptonight-GPU',
       'Algorithm_DPoS', 'Algorithm_Dagger', 'Algorithm_Dagger-Hashimoto',
       'Algorithm_ECC 256K1', 'Algorithm_Equihash',
       'Algorithm_Equihash+Scrypt', 'Algorithm_Ethash', 'Algorithm_Exosis',
       'Algorithm_Green Protocol', 'Algorithm_Groestl', 'Algorithm_HMQ1725',
       'Algorithm_HybridScryptHash256', 'Algorithm_IMesh',
       'Algorithm_Jump Consistent Hash', 'Algorithm_Keccak',
       'Algorithm_Leased POS', 'Algorithm_Lyra2RE', 'Algorithm_Lyra2REv2',
       'Algorithm_Lyra2Z', 'Algorithm_M7 POW', 'Algorithm_Multiple',
       'Algorithm_NIST5', 'Algor

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_data = scaler.fit_transform(crypto_6[["TotalCoinsMined", "TotalCoinSupply"]])
scaled_data = pd.DataFrame(scaled_data, columns=crypto_6.columns[:2])
scaled_data

TotalCoinsMined  TotalCoinSupply
0          -0.117108        -0.152870
1          -0.093970        -0.145009
2           0.524946         4.489424
3          -0.116715        -0.152560
4          -0.114747        -0.152870
..               ...              ...
527        -0.073251        -0.123317
528        -0.116781        -0.149176
529        -0.095613        -0.132179
530        -0.116948        -0.152560
531        -0.117105        -0.152856

[532 rows x 2 columns]

In [12]:
crypto_7 = crypto_6[crypto_6.columns[2:]]
crypto_7 = crypto_7.reset_index()
crypto_7

index  Algorithm_1GB AES Pattern Search  Algorithm_536  \
0        0                                 0              0   
1        2                                 0              0   
2        5                                 0              0   
3        7                                 0              0   
4        8                                 0              0   
..     ...                               ...            ...   
527   1238                                 0              0   
528   1242                                 0              0   
529   1245                                 0              0   
530   1246                                 0              0   
531   1247                                 0              0   

     Algorithm_Argon2d  Algorithm_BLAKE256  Algorithm_Blake  \
0                    0                   0                0   
1                    0                   0                0   
2                    0                   0                0   
3                    0                   0                0   
4                    0                   0                0   
..                 ...                 ...              ...   
527                  0                   0                0   
528                  0                   0                0   
529                  0                   0                0   
530                  0                   0                0   
531                  0                   0                0   

     Algorithm_Blake2S  Algorithm_Blake2b  Algorithm_C11  \
0                    0                  0              0   
1                    0                  0              0   
2                    0                  0              0   
3                    0                  0              0   
4                    0                  0              0   
..                 ...                ...            ...   
527                  0                  0              0   
528                  0                  0              0   
529                  0                  0              0   
530                  0                  0              0   
531                  0                  0              0   

     Algorithm_Cloverhash  ...  ProofType_PoW/PoS  ProofType_PoW/PoS   \
0                       0  ...                  1                   0   
1                       0  ...                  1                   0   
2                       0  ...                  1                   0   
3                       0  ...                  0                   0   
4                       0  ...                  0                   0   
..                    ...  ...                ...                 ...   
527                     0  ...                  0                   0   
528                     0  ...                  1                   0   
529                     0  ...                  0                   0   
530                     0  ...                  0                   0   
531                     0  ...                  0                   0   

     ProofType_PoW/PoW  ProofType_PoW/nPoS  ProofType_Pos  \
0                    0                   0              0   
1                    0                   0              0   
2                    0                   0              0   
3                    0                   0              0   
4                    0                   0              0   
..                 ...                 ...            ...   
527                  0                   0              0   
528                  0                   0              0   
529                  0                   0              0   
530                  0                   0              0   
531                  0                   0              0   

     ProofType_Proof of Authority  ProofType_Proof of Trust  ProofType_TPoS  \
0                               0                         0               0   
1 

In [13]:
crypto_8 = crypto_7.drop(columns=['index'])
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoS  ProofType_PoW/PoS   \
0                         0  ...                  1                   0   
1                         0  ...                  1                   0   
2                         0  ...                  1                   0   
3                         0  ...                  0                   0   
4                         0  ...                  0                   0   
..                      ...  ...                ...                 ...   
527                       0  ...                  0                   0   
528                       0  ...                  1                   0   
529                       0  ...                  0                   0   
530                       0  ...                  0                   0   
531                       0  ...                  0                   0   

     ProofType_PoW/PoW  ProofType_PoW/nPoS  ProofType_Pos  \
0                    0                   0              0   
1                    0                   0              0   
2                    0                   0              0   
3                    0                   0              0   
4                    0                   0              0   
..                 ...                 ...            ...   
527                  0                   0              0   
528                  0                   0              0   
529                  0                   0              0   
530                  0                   0              0   
531               

In [14]:
crypto_8["TotalCoinSupply"] = scaled_data["TotalCoinSupply"]
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoS   ProofType_PoW/PoW  \
0                         0  ...                   0                  0   
1                         0  ...                   0                  0   
2                         0  ...                   0                  0   
3                         0  ...                   0                  0   
4                         0  ...                   0                  0   
..                      ...  ...                 ...                ...   
527                       0  ...                   0                  0   
528                       0  ...                   0                  0   
529                       0  ...                   0                  0   
530                       0  ...                   0                  0   
531                       0  ...                   0                  0   

     ProofType_PoW/nPoS  ProofType_Pos  ProofType_Proof of Authority  \
0                     0              0                             0   
1                     0              0                             0   
2                     0              0                             0   
3                     0              0                             0   
4                     0              0                             0   
..                  ...            ...                           ...   
527                   0              0                             0   
528                   0              0                             0   
529                   0              0   

In [15]:
crypto_8["TotalCoinsMined"] = scaled_data["TotalCoinsMined"]
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoW  ProofType_PoW/nPoS  \
0                         0  ...                  0                   0   
1                         0  ...                  0                   0   
2                         0  ...                  0                   0   
3                         0  ...                  0                   0   
4                         0  ...                  0                   0   
..                      ...  ...                ...                 ...   
527                       0  ...                  0                   0   
528                       0  ...                  0                   0   
529                       0  ...                  0                   0   
530                       0  ...                  0                   0   
531                       0  ...                  0                   0   

     ProofType_Pos  ProofType_Proof of Authority  ProofType_Proof of Trust  \
0                0                             0                         0   
1                0                             0                         0   
2                0                             0                         0   
3                0                             0                         0   
4                0                             0                         0   
..             ...                           ...                       ...   
527              0                             0                         0   
528              0                             0                 

### Dimensionality Reduction

In [16]:
# # Initialize PCA model
# pca = PCA(n_components=0.90)

# # Get two principal components for the data.
# crypto_pca = pca.fit_transform(df_shopping)
